File: phi_plugin_koishi/model/getRksRank.ts
// import { redis } from 'yunzai'

import { redis } from "../components/redis"
import { redisPath } from "./constNum"
import { $ } from 'koishi'


export default new class getRksRank {
    /**
     * 添加成绩
     * @param {string} sessionToken 
     * @param {number} rks 
     */
    async addUserRks(sessionToken: string, rks: number) {
        return await redis.upsert('phigrosUserRks', [{ token: sessionToken, rks: rks }], 'token')
    }

    /**
     * 删除成绩
     * @param {string} sessionToken 
     */
    async delUserRks(sessionToken: string) {
        return await redis.remove('phigrosUserRks', { token: sessionToken })
    }

    /**
     * 获取用户排名
     * @param {number} rks 
     */
    async getUserRank(rks: number): Promise<number> {
        return await redis
            .select('phigrosUserRks')
            .where(row => $.gt(row.rks, rks))
            .execute(row => $.count(row.id))
    }

    /**
     * 获取sessionToken rks
     * @param {number} sessionToken 
     * @returns {Promise<number>}
     */
    async getUserRks(sessionToken: string) {
        return (await redis.get('phigrosUserRks', { token: sessionToken }, ['rks']))[0]
    }

    /**
     * 获取排名
     * @param {number} min 0起
     * @param {number} max 不包含
     */
    async getRankUser(min: number, max: number) {
        return await redis
            .select('phigrosUserRks')
            .orderBy('rks', 'desc')
            .limit(min)
            .offset(max - min)
            .execute()
    }

    /**
     * 获取所有排名
     */
    async getAllRank() {
        return await redis
            .select('phigrosUserRks')
            .orderBy('rks', 'desc')
            .execute()
    }
}()

File: phi_plugin/model/getRksRank.py
# // import { redis } from 'yunzai'

# import { redis } from "../components/redis"
# import { redisPath } from "./constNum"
# import { $ } from 'koishi'
from ..models import phigrosUserRks


class getRksRank:
    @staticmethod
    async def addUserRks(sessionToken: str, rks: float) -> bool:
        """
        添加成绩

        :param sessionToken: SessionToken
        :param rks: rks
        """
        return await phigrosUserRks.set_user_rks(sessionToken, rks)

    @staticmethod
    async def delUserRks(sessionToken: str) -> bool:
        """
        删除成绩

        :param sessionToken: SessionToken
        """
        return await phigrosUserRks.delete_user_rks(sessionToken)

    @staticmethod
    async def getUserRank(sessionToken: str) -> int | None:
        """
        获取用户排名

        :param sessionToken: SessionToken
        """
        return await phigrosUserRks.getUserRank(sessionToken)

    @staticmethod
    async def getUserRks(sessionToken: str) -> float | None:
        """
        获取用户RKS

        :param sessionToken: SessionToken
        """
        return await phigrosUserRks.get_user_rks(sessionToken)

    @staticmethod
    async def getRankUser(min_rank: int, max_rank: int):
        """
        获取指定排名范围的用户数据（基于RKS分数降序）

        :param min_rank: 起始排名（0起始）
        :param max_rank: 结束排名（不包含）
        :return: 包含用户数据的字典列表
        """
        return await phigrosUserRks.getRankUsers(min_rank, max_rank)

    @staticmethod
    async def getAllRank():
        """
        获取所有用户数据（基于RKS分数降序）

        :return: 包含所有用户数据的字典列表
        """
        return await phigrosUserRks.getAllRank()
